### PR TITLE
Fix #667; early stop and export warning

### LINKED ISSUE #667: early stop and export warning
I'm not sure I'm doing correct or wrong, but doing the training with tedlium corpus, I've got the following logs:
```
+ '[' '!' -f DeepSpeech.py ']'
+ '[' '!' -d '' ']'
+ COMPUTE_DATA_DIR=data/ted
+ '[' '!' -f data/ted/ted-train.csv ']'
+ '[' -d '' ']'
+ checkpoint_dir=ckpt/ted
+ python3 -u DeepSpeech.py --train_files data/ted/ted-train.csv --dev_files data/ted/ted-dev.csv --test_files data/ted/ted-test.csv --train_batch_size 12 --dev_batch_size 8 --test_batch_size 8 --epoch 30 --display_step 1 --validation_step
 1 --dropout_rate 0.30 --default_stddev 0.046875 --learning_rate 0.0001 --checkpoint_dir ckpt/ted --export_dir model/ted
I STARTING Optimization
I Training of Epoch 0 - WER: 0.832313, loss: 163.162001103, mean edit distance: 0.428546
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.373605, mean edit distance: 1.000000
I  - src: "i"
I  - res: ""
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.463489, mean edit distance: 0.333333
I  - src: "the"
I  - res: "he"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.628319, mean edit distance: 0.333333
I  - src: "the"
I  - res: "he"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.672621, mean edit distance: 0.500000
I  - src: "so"
I  - res: "o"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.732766, mean edit distance: 0.333333
I  - src: "the"
I  - res: "he"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.741268, mean edit distance: 0.333333
I  - src: "and"
I  - res: "ad"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.750401, mean edit distance: 0.500000
I  - src: "so"
I  - res: "o"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.782803, mean edit distance: 0.500000
I  - src: "so"
I  - res: "o"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.818142, mean edit distance: 0.666667
I  - src: "and"
I  - res: "d"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.921565, mean edit distance: 0.500000
I  - src: "so"
I  - res: "o"
I --------------------------------------------------------------------------------
I Validation of Epoch 0 - WER: 0.592595, loss: 150.599370435, mean edit distance: 0.285013
I --------------------------------------------------------------------------------
I WER: 0.142857, loss: 10.629977, mean edit distance: 0.035714
I  - src: "he had no idea what that was"
I  - res: "ye had no idea what that was"
I --------------------------------------------------------------------------------
I WER: 0.142857, loss: 11.727612, mean edit distance: 0.071429
I  - src: "he had no idea what that was"
I  - res: "he h no idea what that was"
I --------------------------------------------------------------------------------
I WER: 0.200000, loss: 7.491432, mean edit distance: 0.150000
I  - src: "this is the last one"
I  - res: "this is last one"
I --------------------------------------------------------------------------------
I WER: 0.250000, loss: 10.765830, mean edit distance: 0.117647
I  - src: "cause if they did"
I  - res: "as if they did"
I --------------------------------------------------------------------------------
I WER: 0.250000, loss: 11.774858, mean edit distance: 0.176471
I  - src: "cause if they did"
I  - res: "cut if they did"
I --------------------------------------------------------------------------------
I WER: 0.250000, loss: 11.917225, mean edit distance: 0.210526
I  - src: "thank you very much"
I  - res: "ask you very much"
I --------------------------------------------------------------------------------
I WER: 0.285714, loss: 10.644833, mean edit distance: 0.071429
I  - src: "he had no idea what that was"
I  - res: "the ad no idea what that was"
I --------------------------------------------------------------------------------
I WER: 0.500000, loss: 10.977041, mean edit distance: 0.210526
I  - src: "thank you very much"
I  - res: "tank you been much"
I --------------------------------------------------------------------------------
I WER: 0.666667, loss: 10.611268, mean edit distance: 0.250000
I  - src: "all the pictures"
I  - res: "al the picnics"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 9.176581, mean edit distance: 0.444444
I  - src: "thank you"
I  - res: "things"
I --------------------------------------------------------------------------------
I Early stop triggered!! Validation_loss:148.476303 std_of_loss(n_step) :8.185158 mean_loss(n_step): 124.039864
I FINISHED Optimization - training time: 11:50:07
I Test of Epoch 1 - WER: 0.631382, loss: 104.755890629, mean edit distance: 0.310609
I --------------------------------------------------------------------------------
I WER: 0.250000, loss: 5.117770, mean edit distance: 0.105263
I  - src: "if this doesnt work"
I  - res: "it this doesnt work"
I --------------------------------------------------------------------------------
I WER: 0.250000, loss: 6.721038, mean edit distance: 0.157895
I  - src: "if this doesnt work"
I  - res: "if this doesnt or"
I --------------------------------------------------------------------------------
I WER: 0.250000, loss: 6.862726, mean edit distance: 0.105263
I  - src: "if this doesnt work"
I  - res: "if this doesnt or"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 6.740873, mean edit distance: 0.037037
I  - src: "and not just for those fish"
I  - res: "and not just porthole fish"
I --------------------------------------------------------------------------------
I WER: 0.500000, loss: 7.249999, mean edit distance: 0.285714
I  - src: "sort of"
I  - res: "sort"
I --------------------------------------------------------------------------------
I WER: 0.666667, loss: 6.996890, mean edit distance: 0.166667
I  - src: "one billion people"
I  - res: "one bill ion people"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.529166, mean edit distance: 0.500000
I  - src: "so"
I  - res: "to"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 5.450099, mean edit distance: 0.250000
I  - src: "okay"
I  - res: "may"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 5.559149, mean edit distance: 0.250000
I  - src: "okay"
I  - res: "may"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 7.324883, mean edit distance: 0.066667
I  - src: "not intensively"
I  - res: "not intensive by"
I --------------------------------------------------------------------------------
I Exporting the model...
I Restored checkpoint at training epoch 9867
WARNING:tensorflow:From DeepSpeech.py:1634: generic_signature (from tensorflow.contrib.session_bundle.exporter) is deprecated and will be removed after 2017-06-30.
Instructions for updating:
Please use SavedModel instead.
WARNING:tensorflow:From DeepSpeech.py:1636: generic_signature (from tensorflow.contrib.session_bundle.exporter) is deprecated and will be removed after 2017-06-30.
Instructions for updating:
Please use SavedModel instead.
WARNING:tensorflow:From DeepSpeech.py:1636: Exporter.init (from tensorflow.contrib.session_bundle.exporter) is deprecated and will be removed after 2017-06-30.
Instructions for updating:
Please use SavedModel instead.
WARNING:tensorflow:From DeepSpeech.py:1644: Exporter.export (from tensorflow.contrib.session_bundle.exporter) is deprecated and will be removed after 2017-06-30.
Instructions for updating:
Please use SavedModel instead.
Converted 14 variables to const ops.
328 ops in the final graph.
I Models exported at model/ted
```

1. why was the early stop triggered just after one epoch?
2. is it okay about the exporter warning? (I'm using tensorflow 1.2)
3. as testing with other data not including in the corpus, the result shows as like:
```
deepspeech ../model/ted/output_graph.pb conan1.wav 
2017-06-28 15:45:13.095039: I tensorflow/core/common_runtime/gpu/gpu_device.cc:938] Found device 0 with properties: 
name: GeForce GTX 1070
major: 6 minor: 1 memoryClockRate (GHz) 1.683
pciBusID 0000:83:00.0
Total memory: 7.92GiB
Free memory: 7.84GiB
2017-06-28 15:45:13.095116: I tensorflow/core/common_runtime/gpu/gpu_device.cc:959] DMA: 0 
2017-06-28 15:45:13.095125: I tensorflow/core/common_runtime/gpu/gpu_device.cc:969] 0:   Y 
2017-06-28 15:45:13.095140: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1028] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 1070, pci bus id: 0000:83:00.0)
iv e livi in thas sans lis frtu years and i never ben this coled in my life
the original utterance is:
I’ve been living in Los Angeles for two years, and I’ve never been this cold in my life
```

so, it seems the char based output states are not sufficiently trained by the training. Do I need to add some char-based language model? or is it okay to continue with --early_stop false option?

### PR DESCRIPTION
**Pull request by andi4191**
@tilmankamp , @kdavis-mozilla , @reuben : Fix for early stopping trigger after 1st epoch. Turns out validation losses were collected for all the jobs from the single epoch as different epochs on when running on GPUs. Now they are accumulated epoch wise and then considered for decision based on previous 'earlystop_nsteps' epochs.

### CODE REVIEW

**Comment by kdavis-mozilla** [comment]
@andi4191 Doesn't look like you rebased over the latest master; could you do that please?

**Comment by andi4191** [comment]
@kdavis-mozilla : It is showing tilman's latest commit in this branch. I verified from master as well.


commit c8cd6aec90193ff33b7ace542172d41731fa52fa
**Author: andi4191 <a.dixit91@gmail.com>
Date:   Mon Jul 3 21:06:51 2017 -0700

    Fix #667; early stop and export warning

commit 449adf3af8f55ce1980a961111dd412ba8dd2de0
Merge: d832a0d 750d4e8
Author: Tilman Kamp <tilmankamp@posteo.de>
Date:   Mon Jul 3 12:05:41 2017 +0200

    Merge pull request #681 from pecastro/issue668

    Fix issue #668. Hang worker in Start batch.

commit 750d4e89a5c17021dea8b15bcd3c19f9e2e1a9dd
Author: Paulo E. Castro <pauloedgarcastro@gmail.com>
Date:   Sat Jul 1 23:10:23 2017 +0100

    Fix issue #668. Hang worker in Start batch.**

File: DeepSpeech.py [hunk 1]
```diff
@@ -1564,7 +1564,7 @@ def end(self, session):
 
                     # if the job was for validation dataset then append it to the COORD's _loss for early stop verification
                     if (job.set_name == 'dev') and (FLAGS.early_stop is True):
-                        COORD._dev_losses.append(job.loss)
+                        COORD._dev_losses[COORD._epoch-1].append(job.loss)
```
Comment (tilmankamp) [inline DeepSpeech.py]: Yes, this was already wrong in the former PR for this problem. I've overseen it in the first one's review.
However: This fix is too complicated.
The loss is persisted in each validating epoch - look for 
```
self.loss = agg_loss / num_jobs
```
I think this can be used for the early stopping.

**Comment by kdavis-mozilla** [comment]
@andi4191 So you are reverting 449adf3 on purpose?

**Comment by andi4191** [comment]
@kdavis-mozilla : no that wasn't intended. may be while rebasing with master that happened. Corrected that and incorporate the comments from tilman

File: DeepSpeech.py [hunk 2]
```diff
@@ -1170,17 +1174,17 @@ def _next_epoch(self):
...
 
             # Calculate the mean of losses for past epochs
-            mean_loss = np.mean(self._dev_losses[-FLAGS.earlystop_nsteps:-2])
+            mean_loss = np.mean(self._dev_losses[-FLAGS.earlystop_nsteps:-1])
             # Calculate the standard deviation for losses from validation part in the past epochs
-            std_loss = np.std(self._dev_losses[-FLAGS.earlystop_nsteps:-2])
+            std_loss = np.std(self._dev_losses[-FLAGS.earlystop_nsteps:-1])
             # Update the list of losses incurred
             self._dev_losses = self._dev_losses[-FLAGS.earlystop_nsteps:]
             log_debug('Current validation loss: %f, std_of_loss(n_step) :%f mean_loss(n_step): %f' % (self._dev_losses[-1], std_loss, mean_loss))
```
Comment (tilmankamp) [inline DeepSpeech.py]: Better: "Checking for early stopping (last %d steps): loss: %f, standard deviation: %f, mean: %f...".

File: DeepSpeech.py [hunk 3]
```diff
@@ -969,6 +969,10 @@ def done(self):
 
                 self.loss = agg_loss / num_jobs
 
+                # if the job was for validation dataset then append it to the COORD's _loss for early stop verification
+                if (FLAGS.early_stop is True) and (self.set_name == 'dev'):
+                    COORD._dev_losses.append(self.loss)
```
Comment (tilmankamp) [inline DeepSpeech.py]: Much better now - albeit you could also access loss values of the entries in `COORD._epochs_done` by some list comprehension. Your choice.

File: DeepSpeech.py [hunk 4]
```diff
@@ -1170,17 +1174,17 @@ def _next_epoch(self):
...
             # Update the list of losses incurred
             self._dev_losses = self._dev_losses[-FLAGS.earlystop_nsteps:]
             log_debug('Current validation loss: %f, std_of_loss(n_step) :%f mean_loss(n_step): %f' % (self._dev_losses[-1], std_loss, mean_loss))
 
             # Making sure slight fluctuations don't bother the early stopping from performing
-            if self._dev_losses[-1] > np.max(self._dev_losses[:-2]) or (abs(self._dev_losses[-1] - mean_loss) < FLAGS.estop_mean_thresh and std_loss < FLAGS.estop_std_thresh):
+            if self._dev_losses[-1] > np.max(self._dev_losses[:-1]) or (abs(self._dev_losses[-1] - mean_loss) < FLAGS.estop_mean_thresh and std_loss < FLAGS.estop_std_thresh):
                 # Time to early stop
-                log_info('Early stop triggered!! Validation_loss:%f std_of_loss(n_step) :%f mean_loss(n_step): %f' % ((self._dev_losses[-1], std_loss, mean_loss)))
+                log_info('Early stop triggered, validation_loss: %f std_of_loss(n_step): %f mean_loss(n_step): %f' % ((self._dev_losses[-1], std_loss, mean_loss)))
```
Comment (tilmankamp) [inline DeepSpeech.py]: Here you could reason the early stop by having two different messages for the two `or` parts of this `if` block (by making it an `elif`).
E.g. "Early stop, as ..."
Please don't use underscores and abbreviations.

File: DeepSpeech.py (hunk 3 above)
Comment (andi4191) [inline DeepSpeech.py]: Would prefer to stick to a simple design

**Comment by andi4191** [comment]
@tilmankamp : Incorporated comments

File: DeepSpeech.py [hunk 5]
```diff
@@ -1170,17 +1174,17 @@ def _next_epoch(self):
...
+            mean_loss = np.mean(self._dev_losses[-FLAGS.earlystop_nsteps:-1])
             # Calculate the standard deviation for losses from validation part in the past epochs
-            std_loss = np.std(self._dev_losses[-FLAGS.earlystop_nsteps:-2])
+            std_loss = np.std(self._dev_losses[-FLAGS.earlystop_nsteps:-1])
             # Update the list of losses incurred
             self._dev_losses = self._dev_losses[-FLAGS.earlystop_nsteps:]
-            log_debug('Current validation loss: %f, std_of_loss(n_step) :%f mean_loss(n_step): %f' % (self._dev_losses[-1], std_loss, mean_loss))
+            log_debug('Checking for early stopping (last %d steps) validation loss: %f, with standard deviation: %f and mean: %f' % (FLAGS.earlystop_nsteps, self._dev_losses[-1], std_loss, mean_loss))
 
             # Making sure slight fluctuations don't bother the early stopping from performing
```
Comment (reuben) [inline DeepSpeech.py]: Maybe expand this comment clarifying the two different things we're doing here? Something like:

```python
            # Check if the validation loss has started increasing, or is not
            # decreasing substantially, making sure slight fluctuations don't
            # bother the early stopping from working.
```

Comment (andi4191) [inline DeepSpeech.py]: ok
